doc: add package.json documentation

aduh95 opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

Node.js has recently started to rely more on the `package.json` file, and I feel it would be helpful to have a separate document defining the fields we use (a bit like the guide on [npm docs](https://docs.npmjs.com/creating-a-package-json-file)).

Currently, Conditional Exports is supported by both ESM and CJS loaders, and it is documented in the `esm.md` file, which arguably is confusing.

I have used `package.json.md` for the name of the file, maybe we could use something more broad like `package_scope.md` or `packages.md` so it might be useful for other things we want to document. Please weight in if you have thoughts.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

benjamingr commented on doc/api/esm.md:
```diff
@@ -227,204 +167,11 @@ package. It is not a strong encapsulation since a direct require of any
 absolute subpath of the package such as
 `require('/path/to/node_modules/pkg/subpath.js')` will still load `subpath.js`.
 
-#### Subpath Exports
```
Should there be a link from here or some other indication that users should seek content related to ESM exports in the Package docs?

jasnell commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,307 @@
+down until the next folder containing another `package.json`, is considered a
+_package scope_. This document aims to describe the fields used by Node.js
+runtime.
+
```
It would likely be good to say something in here about how the `package.json` file is used by multiple tools and that additional fields may be used that Node.js does not necessarily support.

mcollina replied:
I think it should also link npm documentation about this file.

sam-github replied:
Agree with both comments above, and point to the [yarn docs](https://classic.yarnpkg.com/en/docs/package-json/ ), too.

ljharb replied:
I don't think node docs should be linking to a package manager it doesn't ship; there's far more than npm or yarn out there (there's more than one set of relevant yarn docs, even - v1 vs v2), and it's a slippery slope to include this.

sam-github replied:
@ljharb Even if we don't deep link to npm and yarn docs, we have to mention that the file is used by other tools, or we will be flooded by bug reports about the package.json fields that we "failed to document". Actual links have the downside of possibly becoming broken, but they are what is most useful to readers of the docs, and I think our docs should stive to be most useful.

ljharb replied:
Absolutely it should mention that fields can be used by other tools; I just don't think node docs should be enumerating those.

sam-github replied:
Ah, so you think we should link to only npm, because we only ship npm?

GeoffreyBooth replied:
I think we should link to npm's documentation and mention that other tools use `package.json`, and that's it. Npm is special both because Node ships it and that tool either created the `package.json` file/spec or defined most of the fields in it.

ljharb replied:
@sam-github yes.

aduh95 replied:
Resolved via 98454d647ff23deaf1d0226c0de7406fc3f2fa91

jasnell commented on doc/api/package.json.md:
````diff
@@ -0,0 +1,307 @@
+// Loads ./node_modules/es-module-package/src/features/x.js
+```
+
+When using folder mappings, ensure that you do want to expose every
+module inside the subfolder. Any modules which are not public
+should be moved to another folder to retain the encapsulation
+benefits of exports.
````
```suggestion
When using folder mappings, every module contained within the
folder will be exposed. Any modules which are not public
should be moved to another folder to retain the encapsulation
benefits of exports.
```

MylesBorins commented:
@nodejs/modules this significantly refactors the documentation for ESM PTAL

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -50,65 +50,17 @@ or when referenced by `import` statements within ES module code:
 
-A folder containing a `package.json` file, and all subfolders below that folder
-down until the next folder containing another `package.json`, is considered a
-_package scope_. The `"type"` field defines how `.js` files should be treated
```
We still need this first sentence, that defines what a package scope is. I would perhaps keep this paragraph unchanged.

jkrems replied:
Does the whole section on package scope and file extension belong into the `package.json` section? It doesn't seem to be specific to ESM.

aduh95 replied:
I have added links to the definition of package scope in package.json.html,  do you think that is sufficient?

> Does the whole section on package scope and file extension belong into the package.json section? It doesn't seem to be specific to ESM.

It is actually a sub-section of [Enabling](https://nodejs.org/docs/latest/api/esm.html#esm_enabling), and because CJS is default in node, it makes sense to have it documented here IMO.

GeoffreyBooth commented on doc/api/index.md:
```diff
@@ -38,6 +38,7 @@
 * [Modules](modules.html)
 * [Net](net.html)
 * [OS](os.html)
+* [Package.json](package.json.html)
```
```suggestion
* [package.json](package.json.html)
```

aduh95 replied:
Resolved 5a3959c7ee1c97537afab22c4e56b2926a59bbb2

GeoffreyBooth commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,307 @@
+# Package.json
```
```suggestion
# <code>package.json</code>
```

aduh95 replied:
Resolved via 5a3959c7ee1c97537afab22c4e56b2926a59bbb2

GeoffreyBooth commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,307 @@
+
+A folder containing a `package.json` file, and all subfolders below that folder
+down until the next folder containing another `package.json`, is considered a
+_package scope_. This document aims to describe the fields used by Node.js
+runtime.
```
```suggestion
_package scope_. This document aims to describe the fields used by the
Node.js runtime.
```

GeoffreyBooth replied:
The two sentences in this paragraph don't really belong together. The first defines a package scope, the second introduces this document.

The overall order of sections is a bit off, IMO. I would think that `"name"` should come first, and there should be an explanation of how this field is used with `require` and `import`. When one does `require('foo')`, does Node look for a folder named `node_modules/foo` or a package with `"name": "foo"`? What happens when they disagree? Etc. If Node doesn't use the field at all, we shouldn't be documenting the field.

After that I think should be introduction of package scope and `"type"`. Then exports, conditional exports and finally `"main"`.

GeoffreyBooth replied:
Perhaps in the introduction you can list all the fields that Node recognizes, in alphabetical order, with each field linked to its section? That way we can keep a section name like Conditional Exports, but in the introduction you can have something like “`"exports"`: See [Exports](#) and [Conditional Exports](#)”. And then the sections can be in logical order of what makes the most sense for someone reading this document from top to bottom, but there’s still this table of contents to jump to a particular field.

As for the resolution algorithm, I think it differs in CommonJS versus ESM, so it's probably best kept in each of those sections and linked to from here.

aduh95 replied:
> Perhaps in the introduction you can list all the fields that Node recognizes, in alphabetical order, with each field linked to its section?

Wouldn't that be redundant with the TOC?

GeoffreyBooth replied:
Possibly, though there’s a disparity in the section names like Conditional Exports and the fields like `"exports"`. We probably want to provide a way for people to look them up via either.

aduh95 replied:
@GeoffreyBooth I've made some changes, now the TOC look like this:

```markdown
* `package.json`
    * Introduction
    * Supported fields
        * `"name"`
        * `"exports"`
            * Subpath Exports
            * Package Exports Fallbacks
            * Exports Sugar
            * Conditional Exports
            * Nested conditions
        * `"main"`
        * `"type"`
```

Does that work for you?

GeoffreyBooth replied:
Sure, that looks great!

aduh95 replied:
Maybe I'm missing something, but I don't see how `"type"` and `"exports"` are related... `"type"` is atm related to the `"main"` field, but #33086 should change that.

The way I see it, its sole purpose is to set the correct parser for `.js` files, which is different from other fields that affect the resolution algorithm, which is why I put it last.

Trott commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,313 @@
+
+## Introduction
+
+A `package.json` file defines a package scope: its parent directory, and all
```
```suggestion
A `package.json` file defines a package scope. Its parent directory, and all
```

Trott commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,313 @@
+`package.json`, is considered a _package scope_.
+
+This document aims to describe the fields used by the Node.js
+runtime. Other tools (E.G.:
```
```suggestion
runtime. Other tools (such as
```

Trott commented:
@nodejs/documentation

MylesBorins commented:
@aduh95 I'm in the process of rewriting a chunk of the docs on exports in https://github.com/nodejs/node/pull/33074

Just wanted to give you a heads up as it will likely create a conflict if mine lands first

aduh95 commented:
I've rebased to integrate the changes made in #33220.

Thanks to everyone who has reviewed this PR and given feedback. Is there more I can do to have it ready to land? @GeoffreyBooth, would you say I have done a good enough job of addressing your comments?

guybedford commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,345 @@
+as ES modules and `.cjs` files are always treated as CommonJS.
+
+If the `"type"` field is not provided or is using an unsupported value, it is
+treated as `"commonjs"`.
```
Perhaps we can remove this line, and then just include the `"or is using an unsupported value"` in one of the descriptions of this above, since there is quite a bit of repetition here.

GeoffreyBooth commented on doc/api/package.json.md:
````diff
@@ -0,0 +1,345 @@
+
+```json
+{
+    "exports": "./index.js"
````
```suggestion
  "exports": "./index.js"
```

GeoffreyBooth commented:
Rereading all of this again, I'm struck by how much content really should be shared between docs pages. For example anyone reading the ECMAScript Modules page of the docs really should get a section on Conditional Exports, as most ESM package authors will want to know about that, but this PR moves that content into the `package.json` page of the docs—which makes sense in its own way, since that content isn't specific to ECMAScript Modules but rather applies to both CommonJS and ECMAScript Modules.

This is making me think that maybe the best organizational structure isn't three separate pages—(CommonJS) Modules, ECMAScript Modules and `package.json`—but rather just one page, Modules, that includes all this content. What does everyone else think?

aduh95 commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,345 @@
+<!-- YAML
+added:
+  - v13.1.0
+  - v12.16.0
```
This was added based on a user request (#33143). The introduction explicitly state that the documented fields are the one used by Node.js only. Do you suggest to just remove it or to make it clearer that it can be used alongside older versions of Node.js?

aduh95 commented on doc/api/esm.md:
```diff
@@ -70,69 +70,22 @@ or when referenced by `import` statements within ES module code:
-project’s `node_modules` folder contains its own `package.json` file, so each
-project’s dependencies have their own package scopes. A `package.json` lacking a
-`"type"` field is treated as if it contained `"type": "commonjs"`.
+Every package in a project’s `node_modules` folder usually contains its own
+[`package.json`][] file, so each project’s dependencies have their own [package
+scopes][]. A [`package.json`][] lacking a [`"type"`][] field is treated as if it
```
> It's a little weird to jump straight to discussing node_modules subfolders without first introducing the concept of package scope. The only difference between this version and the old version is that the first two sentences have been removed, but those are pretty vital sentences that define what a package scope is.

Here are my thoughts on this topic. I can see 3 ways for handling it:

1. Move the sentences to `package.json.md#Introduction`, and have a link in `esm.md#Package Scope and File Extensions` to the former.
2. Keep the sentences in  `esm.md#Package Scope and File Extensions` and add link(s) in `package.json` when _package scope_ is mentioned.
3. Duplicate the sentences.

The only one that makes sense to me is `1.`, and that's the one I have implemented in this PR. I have in mind that duplication is never a good thing and package scope doesn't belong to ESM. That being said, I'm open to being persuade otherwise.

> Also do we want to linkify every mention of package.json and "type" etc.?

It was simply easier for me to do a search and replace, and it seemed tedious to me to decide which mention deserves a link, and which doesn't.

aduh95 commented on doc/api/esm.md:
```diff
@@ -277,17 +230,16 @@ path `import feature from 'my-mod/feature/index.js`.
 #### Main Entry Point Export
```
I have moved this section to `package.json.md` based on your comment; I agree it was weird to split the topic, and it's not an ESM specific one.

aduh95 commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,313 @@
+precedence in versions of Node.js that support `"exports"`.
+[Conditional Exports][] can also be used within `"exports"` to define different
+package entry points per environment, including whether the package is
+referenced via `require` or via `import`.
```
Good catch! I have added a line to explicitly state that `"exports"` expects file URL starting with `./` and `"main"` expects a path. I also added that `"exports"` takes precedence over `"main"` only when referencing a package by its name. PTAL.

aduh95 commented on doc/api/package.json.md:
```diff
@@ -0,0 +1,313 @@
+
+This field is ignored when the [`"exports"`][] field is provided. The `"main"`
+field is supported in all versions of Node.js, but its capabilities are limited:
+it only defines the main entry point of the package.
```
TIL you can feed an absolute path to `"main"` and Node.js doesn't complain and load the script even if it's in a completely unrelated directory 🤯

Also worth noting that `"main"` takes a path and `"exports"` file URLs. I wasn't sure how to phrase it, so I just added `Its value is interpreted as a path.`. I'm open to better suggestions x)

aduh95 commented:
New changes PTAL:

- I've moved "Self-referencing a package using its name" from `esm.md` to `package.json.md` – it's not a ESM-specific feature.
- I've merged `esm.md#Packages` section into `esm.md#Enabling`. Given the content of the section, I think it makes more sense like this.

On master, the `esm.md` TOC look like this:
- Introduction
- Enabling
  - package.json "type" field
  - Package Scope and File Extensions
  - `--input-type` flag
- Packages
  - Package Entry Points
    - Main Entry Point Export
    - Subpath Exports
    - Package Exports Fallbacks
    - Exports Sugar
    - Conditional Exports
    - Nested conditions
    - Self-referencing a package using its name
  - Dual CommonJS/ES Module Packages
    - Dual Package Hazard
      - Writing Dual Packages While Avoiding or Minimizing Hazards
        - Approach 1: Use an ES Module Wrapper
        - Approach 2: Isolate State
- `import` Specifiers
  - Terminology
    - `data:` Imports
- ...


On this PR it looks like this:
- Introduction
- Enabling
  - Package Scope and File Extensions
  - `--input-type` flag
  - Package Entry Points
  - Dual CommonJS/ES Module Packages
  - Dual Package Hazard
    - Writing Dual Packages While Avoiding or Minimizing Hazards
      - Approach 1: Use an ES Module Wrapper
      - Approach 2: Isolate State
- `import` Specifiers
  - Terminology
    - `data:` Imports
- ...

guybedford commented:
I tend to agree with @geoffreybooth's points about the structure here.

Just wondering, could it be worth treating the package.json page more like a very quick reference / index, that then links to the specific sections in either `modules.md` or `esm.md` with the full information? Sort of like how the CLI flag section works.

Separating `modules.md` from `esm.md` does seem like a somewhat arbitrary distinction but I think that it is fine due to the history here.

GeoffreyBooth commented:
I've given this some thought, and I _do_ think we should merge `modules.md` and `esm.md`. The `package.json` info can be on that merged page. I think the merged page would still have big “CommonJS Modules” and “ECMAScript Modules” sections, with lots of subsections moved from the current `modules.md` and `esm.md` respectively; but the idea is that all content that is _not_ particular to either module system should be outside of those big system-specific sections. And there's a lot of such content! So much so that it's made me more sure that we _should_ merge the pages.

For example, I assume the Source Maps API applies to both module systems, but currently it only lives in `modules.md`; and conversely Package Exports applies to both but currently only lives in `esm.md`. And so on. Below I've written a draft outline of the merged page, and you can see how many sections are common, outside of the scope of either module system. For each section I've noted which former file the section came from, or if it is new.

#### Modules

- Introduction (new, explain that Node.js has two module systems)
	- CommonJS Modules (based on `modules.md` introduction)
	- ECMAScript Modules (based on `esm.md` introduction)

- Using (based on `esm.md` section “Enabling”)
	- `package.json` `"type"` field
	- Package Scope and File Extensions
	- `--input-type` flag

- CommonJS Modules (new; all subsections from `modules.md`)
	- The Module Wrapper
	- The Module Scope
		- `__dirname`
		- `__filename`
		- `exports`
		- `module`
		- `require(id)`
			- `require.cache`
			- `require.extensions`
			- `require.main`
			- `require.resolve(request[, options])`
				- `require.resolve.paths(request)`
	- The `module` Object
		- `module.children`
		- `module.exports`
			- `exports` shortcut
		- `module.filename`
		- `module.id`
		- `module.loaded`
		- `module.parent`
		- `module.paths`
		- `module.require(id)`
	- Accessing The Main Module
	- Core Modules
	- File Modules
		- `require` of Files with the `.mjs` Extension (from `modules.md` “Addenda: The `.mjs` extension”)
	- Folders as Modules
	- Loading from the Global Folders
	- Loading from `node_modules` Folders
	- Cycles
	- Caching
		- Module Caching Caveats

- ECMAScript Modules (new; all subsections from `esm.md`)
	- `import` Specifiers
		- Terminology
			- `data:` Imports
	- `import.meta`
	- Differences Between ES Modules and CommonJS
		- Mandatory File Extensions
		- No `NODE_PATH`
		- No `require`, `exports`, `module.exports`, `__filename`, `__dirname`
		- No `require.resolve`
		- No `require.extensions`
		- No `require.cache`
		- URL-based Paths
	- Interoperability with CommonJS
		- `require`
		- `import` statements
		- `import()` expressions
	- CommonJS, JSON, and Native Modules
	- Builtin Modules
	- Experimental JSON Modules
	- Experimental Wasm Modules
	- Experimental Top-Level `await`
	- Experimental Loaders
		- Hooks
			- <code>resolve</code> Hook
			- <code>getFormat</code> Hook
			- <code>getSource</code> Hook
			- <code>transformSource</code> Hook
			- <code>getGlobalPreloadCode</code> Hook
			- <code>dynamicInstantiate</code> Hook
		- Examples
			- HTTPS Loader
			- Transpiler Loader

- Packages (from `esm.md`)
	- The `package.json` File (new section introducing the file itself, with very brief subsections that are nothing more than links to the appropriate sections)
		- `"exports"`
		- `"main"`
		- `"name"`
		- `"type"`
	- Package Entry Points
		- Main Entry Point Export
		- Subpath Exports
		- Package Exports Fallbacks
		- Exports Sugar
		- Conditional Exports
		- Nested conditions
		- Self-referencing a package using its name
	- Dual CommonJS/ES Module Packages
		- Dual Package Hazard
		- Writing Dual Packages While Avoiding or Minimizing Hazards
			- Approach &#35;1: Use an ES Module Wrapper
			- Approach &#35;2: Isolate State

- Utility Methods (from `modules.md` “The `Module` Object”)
	- `module.builtinModules`
	- `module.createRequire(filename)`
	- `module.createRequireFromPath(filename)`
	- `module.syncBuiltinESMExports()`)

- Source Map V3 Support (from `modules.md`)
	- `module.findSourceMap(path[, error])`
	- Class: `module.SourceMap`
		- `new SourceMap(payload)`
		- `sourceMap.payload`
		- `sourceMap.findEntry(lineNumber, columnNumber)`

- Resolution Algorithms (new)
	- CommonJS Resolution Algorithm (from `modules.md` section “All Together...”)
	- ECMAScript Modules Resolution Algorithm (from `esm.md` section “Resolution Algorithm”)
		- Features
		- Resolver Algorithm
		- Customizing ESM specifier resolution algorithm

- Tips for Package Manager Authors (from `modules.md` “Package Manager Tips”)

For reference, here are the current full outlines of each page:

<details>
<summary>Modules</summary>

- Accessing the main module
- Addenda: Package Manager Tips
- Addenda: The `.mjs` extension
- All Together...
- Caching
	- Module Caching Caveats
- Core Modules
- Cycles
- File Modules
- Folders as Modules
- Loading from `node_modules` Folders
- Loading from the global folders
- The module wrapper
- The module scope
	- `__dirname`
	- `__filename`
	- `exports`
	- `module`
	- `require(id)`
		- `require.cache`
		- `require.extensions`
		- `require.main`
		- `require.resolve(request[, options])`
			- `require.resolve.paths(request)`
- The `module` Object
	- `module.children`
	- `module.exports`
		- `exports` shortcut
	- `module.filename`
	- `module.id`
	- `module.loaded`
	- `module.parent`
	- `module.paths`
	- `module.require(id)`
- The `Module` Object
	- `module.builtinModules`
	- `module.createRequire(filename)`
	- `module.createRequireFromPath(filename)`
	- `module.syncBuiltinESMExports()`
- Source Map V3 Support
	- `module.findSourceMap(path[, error])`
	- Class: `module.SourceMap`
		- `new SourceMap(payload)`
		- `sourceMap.payload`
		- `sourceMap.findEntry(lineNumber, columnNumber)`

</details>

<details>
<summary>ECMAScript Modules</summary>

- Introduction
- Enabling
	- `package.json` `"type"` field
	- Package Scope and File Extensions
	- `--input-type` flag
- Packages
	- Package Entry Points
		- Main Entry Point Export
		- Subpath Exports
		- Package Exports Fallbacks
		- Exports Sugar
		- Conditional Exports
		- Nested conditions
		- Self-referencing a package using its name
	- Dual CommonJS/ES Module Packages
		- Dual Package Hazard
		- Writing Dual Packages While Avoiding or Minimizing Hazards
			- Approach \#1: Use an ES Module Wrapper
			- Approach \#2: Isolate State
- `import` Specifiers
	- Terminology
		- `data:` Imports
- `import.meta`
- Differences Between ES Modules and CommonJS
	- Mandatory file extensions
	- No `NODE_PATH`
	- No `require`, `exports`, `module.exports`, `__filename`, `__dirname`
	- No `require.resolve`
	- No `require.extensions`
	- No `require.cache`
	- URL-based paths
- Interoperability with CommonJS
	- `require`
	- `import` statements
	- `import()` expressions
- CommonJS, JSON, and Native Modules
- Builtin modules
- Experimental JSON Modules
- Experimental Wasm Modules
- Experimental Top-Level `await`
- Experimental Loaders
	- Hooks
		- <code>resolve</code> hook
		- <code>getFormat</code> hook
		- <code>getSource</code> hook
		- <code>transformSource</code> hook
		- <code>getGlobalPreloadCode</code> hook
		- <code>dynamicInstantiate</code> hook
	- Examples
		- HTTPS loader
		- Transpiler loader
- Resolution Algorithm
	- Features
	- Resolver Algorithm
	- Customizing ESM specifier resolution algorithm

</details>

aduh95 commented:
The changes in this PR has evolved in very different direction since it's been opened, I'm closing it to open a new one, so we can start the conversation fresh.
